### ISSUE TITLE
Go to Definition for Non JS/TS File Imports

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#15146]
From https://github.com/Microsoft/vscode/issues/24276

**Issue**
Users of tools such as webpack often load non js/ts assets, such as css or images, using `require` or `import`:

```js
require('./main.css');
import './image.png'
...
```

Operations such as `go to definition` currently do not work on these paths

**Proposal**
`go to definition` on these resource paths should try to open the corresponding resource in the editor.

**Comment by aluanhaddad** [comment]
They're a couple of great VS Code plugins that do this at the moment.
I assume they work by returning a "definition" that maps to a file and that VS Code then seems adds this as a possible definition to navigate to (I don't really know if this is how it works, but this is how it appears).

Because of this, they suffer from the following problem:

Say I have a declaration file, let's call it `ambiance.d.ts` with the following content
```ts
declare module '*.css' { export default '' as string }
declare module '*.scss' { export default '' as string }
declare module '*.html' { export default '' as string }
```
in this case, despite of the plugins efforts, the user always sees a list of two definitions to go to instead of being able to jump directly to the file.

That is really the only drawback to the current plugins.

Having said that, I always thought it would be nice if TypeScript could validate the existence of these asset files but i think that ship has long ago sailed.

**Comment by mhegazy** [comment]
My main concern is that this is not ES6 compliant. ES6 does not talk about importing non-code modules. So this code is not really valid ES6 and will not work with a native ES6 module implementation.

So we can pretend that it will work, and just let it pass. or leave it the way it is and be spec-compliant.

**Comment by mjbvz** [comment]
I'm OK with not including this as builtin TS functionality given the compliance concerns. Is there any way we can leverage TSServer for these scenarios however? Perhaps with a TSServer plugin?

**Comment by mhegazy** [comment]
This is part of the module resolution logic, so we have to add it in the core.

**Comment by thw0rted** [comment]
Since this is specific to Webpack's implementation of `require()`, I would argue that it should go in `@types/webpack-env`, not core TypeScript.  Isn't that how this is supposed to work?

**Comment by aluanhaddad** [comment]
@thw0rted how is this specific to Webpack?

**Comment by thw0rted** [comment]
I should have said, since it's specific to the loader / bundler's implementation of `require()`.  I was under the impression that TS didn't provide a native capability to require text resources, but maybe I misunderstood.

**Comment by aluanhaddad** [comment]
@thw0rted 
> I should have said, since it's specific to the loader / bundler's implementation of require(). I was under the impression that TS didn't provide a native capability to require text resources, but maybe I misunderstood.

It is correct that TypeScript does not provide this capability, but Webpack is just one of many tools that do.

**Comment by thw0rted** [comment]
I think my point stands, in that case -- nothing in core TypeScript provides for `require("foo.png")` (or css or html or whatever) so it shouldn't be in an ambient declaration unless I've specifically added typings for an environment that does provide that ability.  Right?

**Comment by aluanhaddad** [comment]
Yes you are quite correct. I didn't mean to suggest otherwise I was just mentioning that the notion is more general than Webpack.

**Comment by jpike88** [comment]
Additionally TypeScript should be checking the path of EVERY import that occurs and raise an error. A hacky workaround silences the error, but an incorrect relative path doesn't flag an error.

https://github.com/Microsoft/vscode/issues/48025#issuecomment-382623046

I don't get the arguments regarding strict ES6 spec compliance. This is something webpack, fusebox and pretty much every bundler out there allows and encourages. Widespread real-world use cases matter, so the meaning of what is and isn't 'compliant' needs to be re-evaluated in this case.

**Comment by jpike88** [comment]
Because Typescript doesn't need to load the file contents of non-code imports, a simple path check would be trivial to implement... otherwise every person that uses Typescript + a bundler would need an extra plugin for smooth usage. Considering how ubiquitous bundlers have become, I still think this would be a trivial addition to the Typescript functionality set...

**Comment by octref** [comment]
Just a quick note here: I implemented CSS/SCSS's `@import` using `DocumentLink` as compared to using `Definition` (both can be cmd-clicked for going to another location). I believe it should be possible to add a ts server plugin that generates those `DocumentLink` for text documents.

**Comment by dwelle** [comment]
Note that this doesn't work even for importing `json` files which are AFAIK in es6 module spec.

**Comment by mjbvz** [comment]
@dwelle See the `resolveJsonModule` setting added in TS 2.9: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-9.html

**Comment by dwelle** [comment]
@mjbvz can't seem to make it work, using this `jsconfig.json`:

```js
{
    "compilerOptions": {
        "rootDir": ".",
        "module": "system",
        "resolveJsonModule": true
    },
    /* ... */
}
```

**Comment by dwelle** [comment]
Ok, it seems that `resolveJsonModule` config requires `module: "es6"` (though the error says it needs `node` module, which isn't valid --- I'd assume a `commonjs` would qualify as `node`, but the json resolution doesn't work with that). But `es6` doesn't work with absolute paths (using the `rootDir: "."`).

So I guess I can either choose between absolute paths, or `json` path resolution, not both. It does make sense since the json resolution is part of es6, not commonjs, but as this issue says, I'm not sure that a text editor should be this pedantic. Better if it just works.

**Comment by sbussard** [comment]
This issue has been bugging me a lot. I'm glad someone else brought it up first.

**Comment by ahaq0** [comment]
Currently, what's the best workaround?

**Comment by Lexicality** [comment]
> I think it's important to address exactly what we're trying to solve.

The problems my team and I currently have are:
- Importing a file that doesn't exist should be a lint error. This is technically possible with ESLint but the plugin is incredibly slow
- The imports from the non-js file should be type checked (this is currently handled via ambient module declarations, but I thought it's worth mentioning)
- Typing an import statement in VSCode should show all the files I've configured to be importable, eg css files, template files and so on
- Pressing F12 on a non-js import should take me to that file in VSCode. 

Would something like a `supplimentalExtensions` config value work for this? It would then entirely be up to the user to make sure the module resolution system is the same between Webpack and Typescript.

(Presumably the typescript compiler would emit an error if asked to bundle a file with a non-js import)

**Comment by joshglenn** [comment]
I had read this issue earlier today, before I filed [issue 90155 in vscode](https://github.com/microsoft/vscode/issues/90155) . This does not appear to me to address my problem. 

But if, as you said earlier in this thread, there are extensions that workaround this I'd sure like to know of one. I've been scouring the extensions gallery trying to find it. 

Do you know any by name that I could try?

**Comment by DMQ** [comment]
I found a vscode plugin called `file peek` to provide this functionality. Hope also can help you guys.

**Comment by Cpanmac** [comment]
When can we deal with this problem？@mjbvz

**Comment by mishannn** [comment]
It is very important to autocomplete files other than JS or TS and check the existence of a file. Due to its absence, typos often appear and, in general, work becomes less comfortable. Please make this opportunity.

**Comment by FrederickEngelhardt** [comment]
Bump. It would be nice to have a way to do file path resolution with wild cards and extrapolate the beginning path and resolve the file using node.

```
declare module '*.css?raw' {
  const content: { [className: string]: string }
  file = require('[filepath].css')

  export default content
  return file
}
```

The above would both declare the typing and return the file it should be resolving. Of course there would need to be escape hatches for bad paths etc.

**Comment by edouard-lopez** [comment]
In the meantine, use can leverage [CSS Navigation extension](https://github.com/pucelle/vscode-css-navigation/) className resolution.

It planned to [support `import` resolution](https://github.com/pucelle/vscode-css-navigation/issues/21#issuecomment-783291250) in a future release.

**Comment by dummdidumm** [comment]
I think this can be closed since this landed in TS 4.3 https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#go-to-def-non-js

**Comment by Yegorich555** [comment]
> I think this can be closed since this landed in TS 4.3 https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#go-to-def-non-js

Yep. I've just rechecked. Looks lite it works now

**Comment by chkpnt** [comment]
Yes, go to definition works. But how can I get rid of the warning _"Cannot find module '../img/bla.png' or its corresponding type declarations.ts(2307)"_ while preserving this feature?

If I `declare module '*.png'` in `global.d.ts` the warning disappears, but then I'm again unable to use go-to-definition.

**Comment by NerdPraise** [comment]
@chkpnt  A workaround; not too good but it works
`declare module '*.png' { export default '' as any; }`

Removes the warning and also go-to-definition

**Comment by chkpnt** [comment]
Unfortunately,  adding `{ export default '' as any; }` doesn't help. When I use go-to-definition, I end up in `global.d.ts` (which might be right, as there is the definition), but I want to open the file I'm importing. 🤷‍♂️

**Comment by Kreozot** [comment]
Any updates on this? Does anybody have a workaround to make go-to-definition work?

**Comment by KokoDoko** [comment]
I have to say I find it quite baffling that importing an image in parcel results in so many problems and pages full of discussions. Isn't it one of the most rudimentary functions of a webpage to show images? Isn't it the function of module bundlers to create a `dist` folder that only has the assets that you actually `import` ? (instead of an ever growing image folder, where you don't know which image is actually used and which image isn't).

**Comment by aleclarson** [comment]
Here's an extension that implements the requested feature:
https://marketplace.visualstudio.com/items?itemName=thisismanta.third-eye

**Comment by appxcess** [comment]
> Here's an extension that implements the requested feature: https://marketplace.visualstudio.com/items?itemName=thisismanta.third-eye

Its not solving the .scss navigation to file and style class.
Issue still exist.

**Comment by shankarThiyagaraajan** [comment]
> Here's an extension that implements the requested feature: https://marketplace.visualstudio.com/items?itemName=thisismanta.third-eye

It’s not solving the '.scss' navigation to file and style class.
Issue still exist.

**Comment by fregante** [comment]
> I think this can be closed since this landed in TS 4.3 [devblogs.microsoft.com/typescript/announcing-typescript-4-3#go-to-def-non-js](https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#go-to-def-non-js)

This seems to work for relative paths (`import "./file.scss"`) but not for node_modules imports:

```js
import "bootstrap-daterangepicker/daterangepicker.css";
import "primeicons/primeicons.css";
import "@my-monorepo/my-lib/stylesheet.scss";
```

**Comment by thw0rted** [comment]
With universal(ish) support for real ESM, do a lot of folks still use these fake resource import / require statements that only work in bundlers? Seems to me that if the bundlers want to let you inject a resource inline, they should come up with a new keyword instead of overloading one that your runtime understands.  (Or, y'know, just get the resource with fetch or fs.read*)
